ios/sound.mm

```diff
@@ -24,209 +24,277 @@
+
+    // copy input data
+    for ( int i = 0; i < pSound->iCoreAudioBufferSizeMono; i++ )
+    {
+        // copy left and right channels separately
+        pSound->vecsTmpAudioSndCrdStereo[2 * i]     = (short) ( pData[2 * i] * _MAXSHORT );     // left
+        pSound->vecsTmpAudioSndCrdStereo[2 * i + 1] = (short) ( pData[2 * i + 1] * _MAXSHORT ); // right
+    }
+    pSound->ProcessCallback ( pSound->vecsTmpAudioSndCrdStereo );
 }
 
-QString CSound::CheckDeviceCapabilities ( const int iDriverIdx )
+/* Implementation *************************************************************/
+CSound::CSound ( void ( *fpNewProcessCallback ) ( CVector<short>& psData, void* arg ),
+                 void*          arg,
+                 const QString& strMIDISetup,
+                 const bool,
+                 const QString& ) :
+    CSoundBase ( "CoreAudio iOS", fpNewProcessCallback, arg, strMIDISetup )
 {
-  /*// This function checks if our required input/output channel
-  // properties are supported by the selected device. If the return
-  // string is empty, the device can be used, otherwise the error
-  // message is returned.
-
-  // check sample rate of input
-  audioSession.setPreferredSampleRate( SYSTEM_SAMPLE_RATE_HZ );
-  // if false, try to set it
-  if ( audioSession.sampleRate != SYSTEM_SAMPLE_RATE_HZ )
-  {
-    throw CGenErr ( tr ("Could not set sample rate. Please close other applications playing audio.") );
-  }
-
-  // special case with 4 input channels: support adding channels
-    if ( ( lNumInChan == 4 ) && bInputChMixingSupported )
+    try
     {
-        // add four mixed channels (i.e. 4 normal, 4 mixed channels)
-        lNumInChanPlusAddChan = 8;
-
-        for ( int iCh = 0; iCh < lNumInChanPlusAddChan; iCh++ )
-        {
-            int iSelCH, iSelAddCH;
-
-            GetSelCHAndAddCH ( iCh, lNumInChan, iSelCH, iSelAddCH );
-
-            if ( iSelAddCH >= 0 )
-            {
-                // for mixed channels, show both audio channel names to be mixed
-                channelInputName[iCh] =
-                    channelInputName[iSelCH] + " + " + channelInputName[iSelAddCH];
-            }
-        }
+        NSError* audioSessionError = nil;
+
+        [[AVAudioSession sharedInstance] setCategory:AVAudioSessionCategoryPlayAndRecord error:&audioSessionError];
+        [[AVAudioSession sharedInstance] requestRecordPermission:^( BOOL granted ) {
+          if ( granted )
+          {
+              // ok
+          }
+        }];
+        [[AVAudioSession sharedInstance] setMode:AVAudioSessionModeMeasurement error:&audioSessionError];
     }
-    else
+    catch ( const CGenErr& generr )
     {
-        // regular case: no mixing input channels used
-        lNumInChanPlusAddChan = lNumInChan;
+        qDebug ( "Sound exception ...." );
     }
-*/
-    // everything is ok, return empty string for "no error" case
-    return "";
+    isInitialized = false;
```

Comment:
Shouldn't this be part of the constructor init right at the top?
```
CSound::CSound ( void ( *fpNewProcessCallback ) ( CVector<short>& psData, void* arg ),
                 void*          arg,
                 const QString& strMIDISetup,
                 const bool,
                 const QString& ) :
    isInitialized ( false ),
    CSoundBase ( "CoreAudio iOS", fpNewProcessCallback, arg, strMIDISetup )
{
...
```

Reply:
I agree a constructor should be at the top. Must it though?

Lol just trolling you. I'll update the code

Reply:
hmm I misunderstood.

The reason was simple: I didn't know how/when to put it in the constructor init. Not a developer at heart.